firebase-firestore/src/main/java/com/google/firebase/firestore/util/ProtoValueUtil.java

```diff
@@ -0,0 +1,266 @@
+      default:
+        return left.equals(right);
+    }
+  }
+
+  private static boolean objectEquals(Value left, Value right) {
+    MapValue leftMap = left.getMapValue();
+    MapValue rightMap = right.getMapValue();
+
+    if (leftMap.getFieldsCount() != rightMap.getFieldsCount()) {
+      return false;
+    }
+
+    for (Map.Entry<String, Value> entry : leftMap.getFieldsMap().entrySet()) {
+      Value otherEntry = rightMap.getFieldsMap().get(entry.getKey());
+      if (!entry.getValue().equals(otherEntry)) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  private static boolean numberEquals(Value left, Value right) {
+    if (left.getValueTypeCase() == Value.ValueTypeCase.INTEGER_VALUE
+        && right.getValueTypeCase() == Value.ValueTypeCase.INTEGER_VALUE) {
+      return left.equals(right);
+    } else if (left.getValueTypeCase() == Value.ValueTypeCase.DOUBLE_VALUE
+        && right.getValueTypeCase() == Value.ValueTypeCase.DOUBLE_VALUE) {
+      return Double.doubleToLongBits(left.getDoubleValue())
+          == Double.doubleToLongBits(right.getDoubleValue());
+    }
+
+    return false;
+  }
+
+  private static boolean arrayEquals(Value left, Value right) {
+    ArrayValue leftArray = left.getArrayValue();
+    ArrayValue rightArray = right.getArrayValue();
+
+    if (leftArray.getValuesCount() != rightArray.getValuesCount()) {
+      return false;
+    }
+
+    for (int i = 0; i < leftArray.getValuesCount(); ++i) {
+      if (!equals(leftArray.getValues(i), rightArray.getValues(i))) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  public static int compare(Value left, Value right) {
+    int leftType = typeOrder(left);
+    int rightType = typeOrder(right);
+
+    if (leftType != rightType) {
+      return Util.compareIntegers(leftType, rightType);
+    }
+
+    switch (leftType) {
+      case FieldValue.TYPE_ORDER_NULL:
+        return 0;
+      case FieldValue.TYPE_ORDER_BOOLEAN:
+        return Util.compareBooleans(left.getBooleanValue(), right.getBooleanValue());
+      case FieldValue.TYPE_ORDER_NUMBER:
+        return compareNumbers(left, right);
+      case FieldValue.TYPE_ORDER_TIMESTAMP:
+        return compareTimestamps(left, right);
+      case FieldValue.TYPE_ORDER_STRING:
+        return left.getStringValue().compareTo(right.getStringValue());
+      case FieldValue.TYPE_ORDER_BLOB:
+        return Util.compareByteString(left.getBytesValue(), right.getBytesValue());
+      case FieldValue.TYPE_ORDER_REFERENCE:
+        return compareReferences(left, right);
+      case FieldValue.TYPE_ORDER_GEOPOINT:
+        return compareGeoPoints(left, right);
+      case FieldValue.TYPE_ORDER_ARRAY:
+        return compareArrays(left, right);
+      case FieldValue.TYPE_ORDER_OBJECT:
+        return compareMaps(left, right);
+      default:
+        throw fail("Invalid value type: " + leftType);
+    }
+  }
+
+  private static int compareMaps(Value left, Value right) {
```

Comment:
Nit: consider ordering these to match their usage above.

Reply:
Done